src/commands/utils/clusters.ts

```diff
@@ -73,45 +74,180 @@ export function getAksClusterSubscriptionItem(commandTarget: any, cloudExplorer:
+    let kubeconfigYaml: any;
+    try {
+        kubeconfigYaml = yaml.load(unauthenticatedKubeconfig.result);
+    } catch (e) {
+        return { succeeded: false, error: `Failed to parse kubeconfig YAML for ${cluster.name}: ${e}` };
+    }
+
+    // We expect there will be just one user in the kubeconfig. Read the 'exec' arguments for it.
+    const execBlock = kubeconfigYaml.users[0].user.exec;
+    const execOptions = readExecOptions(execBlock.args);
+
+    // We need to supply an access token that grants the user access to the cluster. The user running this will be authenticated,
+    // and we have access to their credentials here...
+    const localToken = await cluster.subscription.credentials.getToken();
+
+    // But, the user's token is for VS Code, and instead we need a token whose audience is the AKS server ID.
+    // This can be obtained by exchanging the user's refresh token for one with the new audience.
+    const aadAccessToken = await getAksAadAccessToken(cluster.subscription.environment, execOptions.serverId, execOptions.tenantId, localToken.refreshToken);
+    if (failed(aadAccessToken)) {
+        return aadAccessToken;
+    }
+
+    // Now we have a token the user can use to access their cluster, but kubelogin doesn't support that being passed as an
+    // argument directly. Instead, we can save it to a temp directory and instruct kubelogin to use that as its cache directory.
+    const cacheDir = storeCachedAadToken(aadAccessToken.result, execOptions);
+    if (failed(cacheDir)) {
+        return cacheDir;
+    }
+
+    // This extension controls the version of kubelogin used, so that:
+    // 1. We don't need to rely on the user having previously downloaded it, and
+    // 2. This workflow doesn't get broken by kubelogin behaviour changes between versions
+    const kubeloginPath = await getKubeloginBinaryPath();
+    if (failed(kubeloginPath)) {
+        return kubeloginPath;
+    }
+
+    // Update the kubeconfig YAML with the new kubelogin options, and return the serialized output as a string.
+    execBlock.command = kubeloginPath.result;
+    execBlock.args = buildExecOptionsWithCache(execOptions, cacheDir.result);
+    const authenticatedKubeConfig = yaml.dump(kubeconfigYaml);
+    return { succeeded: true, result: authenticatedKubeConfig };
+}
+
+interface ExecOptions {
+    subcommand: string,
+    environment: string,
+    serverId: string,
+    clientId: string,
+    tenantId: string,
+    loginMethod: string
+}
+
+function readExecOptions(execArgs: [string]): ExecOptions {
+    // The 'exec' command args are made up of a subcommand (get-token) and a number of name/value options.
+    const [subcommand, ...options] = execArgs;
+
+    // Extract the options into an object for convenient lookup.
+    const optionLookup: { [name: string]: string } = options.reduce((result: { [name: string]: string }, _arg: string, index: number) => {
+        return index % 2 === 0 ?
```

Comment:
💡 I wonder if we add couple of comments around this etc. just like details you had in few functions above.

Reply:
Good point, that code isn't obvious at all - hope this explains it a bit better.